contracts/superswap/RexSuperSwap.sol

```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+  /**
+   * @dev Swaps `amountIn` of `_from` SuperToken for at least `amountOutMin`
+   * of `_to` SuperToken through `path` with `poolFees` fees for each pair.
+   *
+   * Returns the amount of `_to` SuperToken received.
+   */
+  function swap(
+    ISuperToken _from,
+    ISuperToken _to,
+    uint256 amountIn,
+    uint256 amountOutMin,
+    address[] memory path,
+    uint24[] memory poolFees, // Example: 0.3% * 10000 = 3000
+    bool _hasUnderlyingFrom,
+    bool _hasUnderlyingTo
+  ) external payable returns (uint256 amountOut) {
+    require(amountIn > 0, "Amount cannot be 0");
+    require(path.length > 1, "Incorrect path");
+    require(poolFees.length == path.length - 1, "Incorrect poolFees length");
+
+    // Step 1: Get underlying tokens and verify path
+    address fromBase = address(_from);
+    address toBase = address(_to);
+    if (_hasUnderlyingFrom) {
+      fromBase = _from.getUnderlyingToken();
+    }
+    if(_hasUnderlyingTo){
+      toBase = _to.getUnderlyingToken();
+    }
+
+    require(path[0] == fromBase, "Invalid 'from' base token");
+    require(path[path.length - 1] == toBase, "Invalid 'to' base token");
+
+    // Step 2: Transfer SuperTokens from sender
+    TransferHelper.safeTransferFrom(
+      address(_from),
+      msg.sender,
+      address(this),
+      amountIn
+    );
+
+    console.log("starting balance of from token - ", _from.balanceOf(address(this)));
+
+    // Step 3: Downgrade if it's not a native token
+    if(_hasUnderlyingFrom){
+      _from.downgrade(amountIn);
+    }
+
+    // Encode the path for swap
+    bytes memory encodedPath;
+    for (uint256 i = 0; i < path.length; i++) {
+      if (i == path.length - 1) {
+        encodedPath = abi.encodePacked(encodedPath, path[i]);
+      } else {
+        encodedPath = abi.encodePacked(encodedPath, path[i], poolFees[i]);
+      }
+    }
+
+    // Approve the router to spend token supplied (fromBase).
+    approve(IERC20(fromBase), address(swapRouter));
+   
+    console.log("balance of from token before swap - ", ERC20(fromBase).balanceOf(address(this)));
+
+    IV3SwapRouter.ExactInputParams memory params = IV3SwapRouter
+      .ExactInputParams({
+        path: encodedPath,
+        recipient: address(this),
+        amountIn: ERC20(fromBase).balanceOf(address(this)),
+        amountOutMinimum: amountOutMin
+      });
+ 
+    // Execute the swap
+    try swapRouter.exactInput(params) returns (uint256 swappedAmount) {
+      amountOut = swappedAmount;
+    } catch Error(string memory reason) {
+      emit ErrorOnSwap(reason);
+    } catch (bytes memory returnData) {
+      emit ReturnDataEvent(returnData);
+    }
+  
+    console.log("balance of to token after swap - ", ERC20(toBase).balanceOf(address(this)));
+  
+    // Step 5: Upgrade and send tokens back
+    approve(IERC20(toBase), address(_to));
+
+    // Upgrade if it's not a native SuperToken
+    if (_hasUnderlyingTo) {
+      if (address(_to) == superNativeToken) {
```

Comment:
@gitChimp88 We don't need this condition, here I wrongly assumed that Super Matic has underlaying token as 0x (native token). As a result the Matic <> Ric test might be failing.